File: Android/app/src/main/java/realm/io/realmpop/controller/game/GameTimer.java
package realm.io.realmpop.controller.game;

import java.util.Date;
import java.util.Locale;
import java.util.Timer;
import java.util.TimerTask;

public class GameTimer {

    private static final long UPDATE_INTERVAL = 100;

    private long expirationInSeconds = 1;
    private Timer timer;
    private Date startedAt = new Date();

    public GameTimer(int expirationInSeconds) {
        this.expirationInSeconds = expirationInSeconds;
    }

    private long timeElapsedInSeconds() {
        return timeElapsedInMillis() / 1000;
    }

    private long timeElapsedInMillis() {
        return new Date().getTime() - startedAt.getTime();
    }

    private String timeElapsedString() {
        double totalTimeElapsedMillis = timeElapsedInMillis();
        double asSeconds = totalTimeElapsedMillis / 1000d;
        return String.format( Locale.US, "%.1f", asSeconds );
    }

    private boolean timeHasExpired() {
        return timeElapsedInSeconds() >= expirationInSeconds;
    }

    private boolean isTimerRunning() {
        return timer != null && startedAt != null;
    }

    public void stopTimer() {
        if(timer != null) {
            timer.cancel();
            timer = null;
            startedAt = null;
        }
    }

    public void startTimer(final TimerDelegate timerDelegate) {
        if(timer != null) {
            stopTimer();
        }
        timer = new Timer();
        startedAt = new Date();
        timer.scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                timerDelegate.onTimerUpdate(new TimerEvent(timeElapsedString()));
                if(timeHasExpired()) {
                    timerDelegate.onTimeExpired(new TimeExpiredEvent(timeElapsedString()));
                    stopTimer();
                }
            }
        }, new Date(), UPDATE_INTERVAL);
    }

    public void expireInSeconds(int endInSeconds) {
        if(isTimerRunning()) {
            long previousExpirationTimeFromStart = expirationInSeconds;
            long newExpirationTimeFromStart = timeElapsedInSeconds() + endInSeconds;
            if(newExpirationTimeFromStart < previousExpirationTimeFromStart) {
                expirationInSeconds = newExpirationTimeFromStart;
            }
        }
    }

    public class TimerEvent {
        private TimerEvent(String timeElapsedString) {
            this.timeElapsedString = timeElapsedString;
        }
        public String timeElapsedString;
    }

    public class TimeExpiredEvent {
        private TimeExpiredEvent(String timeElapsedString){
            this.timeElapsedString = timeElapsedString;
        }
        public String timeElapsedString;
    }

    public interface TimerDelegate {
        void onTimerUpdate(TimerEvent timerEvent);
        void onTimeExpired(TimeExpiredEvent timeExpiredEvent);
    }
}

File: Android/app/src/main/java/realm/io/realmpop/ui/game/GameTimer.kt
package realm.io.realmpop.ui.game

import java.util.Date
import java.util.Locale
import java.util.Timer
import java.util.TimerTask

class GameTimer(var expirationInSeconds: Long = 1L) {

    class TimerEvent(val timeElapsedString: String)
    class TimeExpiredEvent(val timeElapsedString: String)

    interface TimerDelegate {
        fun onTimerUpdate(timerEvent: TimerEvent)
        fun onTimeExpired(timeExpiredEvent: TimeExpiredEvent)
    }

    private val UPDATE_INTERVAL = 100L
    private var timer: Timer? = null
    private var startedAt: Date? = Date()

    private fun timeElapsedInSeconds() = timeElapsedInMillis() / 1000

    private fun timeElapsedInMillis(): Long {
        val localStartedAt = startedAt
        if (localStartedAt != null) {
            return Date().time - startedAt!!.time
        } else {
            return 0
        }
    }

    private fun timeElapsedString(): String {
        val totalTimeElapsedMillis = timeElapsedInMillis().toDouble()
        val asSeconds = totalTimeElapsedMillis / 1000.0
        return String.format(Locale.US, "%.1f", asSeconds)
    }

    private fun timeHasExpired(): Boolean {
        return timeElapsedInSeconds() >= expirationInSeconds
    }

    private fun isTimerRunning()= timer != null && startedAt != null

    fun stopTimer() {
        timer?.cancel()
        timer = null
        startedAt = null
    }

    fun startTimer(timerDelegate: TimerDelegate) {
        if (timer != null) {
            stopTimer()
        }
        timer = Timer()
        startedAt = Date()
        timer!!.scheduleAtFixedRate(object : TimerTask() {
            override fun run() {
                timerDelegate.onTimerUpdate(TimerEvent(timeElapsedString()))
                if (timeHasExpired()) {
                    timerDelegate.onTimeExpired(TimeExpiredEvent(timeElapsedString()))
                    stopTimer()
                }
            }
        }, Date(), UPDATE_INTERVAL)
    }

    fun expireInSeconds(endInSeconds: Int) {
        if (isTimerRunning()) {
            val previousExpirationTimeFromStart = expirationInSeconds
            val newExpirationTimeFromStart = timeElapsedInSeconds() + endInSeconds
            if (newExpirationTimeFromStart < previousExpirationTimeFromStart) {
                expirationInSeconds = newExpirationTimeFromStart
            }
        }
    }

}
